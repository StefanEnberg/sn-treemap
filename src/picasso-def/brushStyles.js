const HIGHLIGHT_STROKE_WIDTH = 2;
export const active = () => ({
  stroke: 'black',
  strokeWidth: (d) => {
 
 
  const {select} = d.data;
  const {source} = d.data;
  return source?.field === select?.source?.field ? HIGHLIGHT_STROKE_WIDTH : 0;
  },
  opacity: (d) => d?.data?.overlayLabel || d?.data?.parentGroup ? 0 : 1,
});

export const inactive = () => ({
  opacity: (d) => (d?.data?.parentGroup ? 0 : 0.5),
});
