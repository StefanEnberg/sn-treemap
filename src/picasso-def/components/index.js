import {treemap} from './treemap';
import {tooltip} from './tooltip';
import {nativeLegend} from './nativeLegend';

export {treemap, tooltip, nativeLegend};
