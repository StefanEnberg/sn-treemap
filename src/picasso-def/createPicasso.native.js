import {createPicasso as carbonCreatePicasso} from '@qlik/carbon-core';

export const createPicasso = (config) => {
  return carbonCreatePicasso(config);
};
