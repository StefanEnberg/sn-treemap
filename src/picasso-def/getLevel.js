export const getLevel = ({qHyperCube}) => {
  const {qDimensionInfo} = qHyperCube;
  let level = qDimensionInfo.reduce((acc, curr, index) => {
    if (qDimensionInfo[index].qStateCounts.qSelected === 1) {
      return acc + 1;
    }
    return acc;
  }, 0);
  level = Math.min(level, qDimensionInfo.length - 1);
  return level;
};

export const getNextSelecteDim = ({qHyperCube}) => {
  // find first dim without a selection
  const {qDimensionInfo} = qHyperCube;
  const dim = qDimensionInfo.findIndex(
    (info) => info.qStateCounts.qSelected === 0,
  );
  return dim === -1 ? qDimensionInfo.length - 1 : dim;
};
