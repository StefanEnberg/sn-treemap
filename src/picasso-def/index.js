import { picassoDef } from './picassoDef';

export { picassoDef };
