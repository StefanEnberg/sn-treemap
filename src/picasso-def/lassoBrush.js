export const lassoBrush = () => ({
  key: 'lasso',
  type: 'brush-lasso',
  settings: {
    brush: {
      components: [
        {
          key: 'treemap',
          data: ['select'],
          contexts: ['lassoContext'],
          action: 'add',
        },
      ],
    },
  },
});
