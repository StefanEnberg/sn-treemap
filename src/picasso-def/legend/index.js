import {legend} from './legend';

export {legend};