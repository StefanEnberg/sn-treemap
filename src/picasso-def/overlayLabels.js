export const overlayLables = () => ({
  type: 'labels',
  key: 'overlay',
  layout: {
    displayOrder: 10,
  },
  settings: {
    sources: [
      {
        component: 'treemap',
        selector: 'rect',
        strategy: {
          type: 'rows',
          settings: {
            fontFamily: 'arial',
            fontSize: 24,
            justify: 0.5,
            align: 0.5,
            padding: 8,
            labels: [
              {
                wordBreak: 'break-word',
                fill: (d) => {
                  if (d?.data?.overlayLabel && d.data.color) {
                    return d.data.color;
                  }
                  return 'rgba(0, 0, 0, 0.5)';
                },
                label: (d) => d?.data?.overlayLabel || '',
              },
            ],
          },
        },
      },
    ],
  },
});
