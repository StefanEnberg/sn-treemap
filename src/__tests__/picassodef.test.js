import { picassoDef } from '../picasso-def/picassoDef';
import testdata from './testdata.json';
import {__DO_NOT_USE__ as NebualInternals}  from '@nebula.js/stardust';
import theme from '@qlik-trial/sense-themes-default/dist/horizon/theme.json';

describe('Treemap picassodef render', () => {
  const {generator, theme: themeFn} = NebualInternals;
  it('should render ok', () => {
    const layout = testdata.qLayout[0].value;
    const _theme = themeFn();
    _theme.internalAPI.setTheme(theme, 'horizon');
    const data = picassoDef({layout, theme:_theme.externalAPI, env: {}});
    expect(data).toMatchSnapshot();
  })
})