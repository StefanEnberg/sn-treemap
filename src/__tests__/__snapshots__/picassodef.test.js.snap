// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Treemap picassodef render should render ok 1`] = `
Object {
  "collections": Array [
    Object {
      "data": Object {
        "hierarchy": Object {
          "label": [Function],
          "props": Object {
            "color": Object {
              "field": "qMeasureInfo/0",
            },
            "expressionColor": undefined,
            "expressionColorText": undefined,
            "select": Object {
              "field": "qDimensionInfo/0",
            },
          },
          "value": [Function],
        },
      },
      "key": "hierarchy",
    },
  ],
  "components": Array [
    Object {
      "brush": Object {
        "consume": Array [
          Object {
            "context": "dataContext",
            "data": Array [
              "select",
            ],
            "filter": [Function],
            "style": Object {
              "active": Object {
                "opacity": [Function],
                "stroke": "black",
                "strokeWidth": [Function],
              },
              "inactive": Object {
                "opacity": [Function],
              },
            },
          },
          Object {
            "context": "lassoContext",
            "data": Array [
              "select",
            ],
            "filter": [Function],
            "style": Object {
              "active": Object {
                "opacity": [Function],
                "stroke": "black",
                "strokeWidth": [Function],
              },
              "inactive": Object {
                "opacity": [Function],
              },
            },
          },
        ],
        "trigger": Array [
          Object {
            "contexts": Array [
              "dataContext",
            ],
            "data": Array [
              "select",
            ],
            "on": "tap",
          },
        ],
      },
      "data": Object {
        "collection": Object {
          "key": "hierarchy",
        },
      },
      "key": "treemap",
      "settings": Object {
        "color": Object {
          "mode": "byMeasure",
        },
        "formatter": Array [
          NumDateTimeFormatter {
            "_dateFormatter": DateFormatter {
              "localeInfo": Object {
                "qCalendarStrings": Object {
                  "qDayNames": Array [
                    "Mon",
                    "Tue",
                    "Wed",
                    "Thu",
                    "Fri",
                    "Sat",
                    "Sun",
                  ],
                  "qLongDayNames": Array [
                    "Monday",
                    "Tuesday",
                    "Wednesday",
                    "Thursday",
                    "Friday",
                    "Saturday",
                    "Sunday",
                  ],
                  "qLongMonthNames": Array [
                    "January",
                    "February",
                    "March",
                    "April",
                    "May",
                    "June",
                    "July",
                    "August",
                    "September",
                    "October",
                    "November",
                    "December",
                  ],
                  "qMonthNames": Array [
                    "Jan",
                    "Feb",
                    "Mar",
                    "Apr",
                    "May",
                    "Jun",
                    "Jul",
                    "Aug",
                    "Sep",
                    "Oct",
                    "Nov",
                    "Dec",
                  ],
                },
              },
              "locale_days": Array [
                "Monday",
                "Tuesday",
                "Wednesday",
                "Thursday",
                "Friday",
                "Saturday",
                "Sunday",
              ],
              "locale_days_abbr": Array [
                "Mon",
                "Tue",
                "Wed",
                "Thu",
                "Fri",
                "Sat",
                "Sun",
              ],
              "locale_months": Array [
                "January",
                "February",
                "March",
                "April",
                "May",
                "June",
                "July",
                "August",
                "September",
                "October",
                "November",
                "December",
              ],
              "locale_months_abbr": Array [
                "Jan",
                "Feb",
                "Mar",
                "Apr",
                "May",
                "Jun",
                "Jul",
                "Aug",
                "Sep",
                "Oct",
                "Nov",
                "Dec",
              ],
              "pattern": "0.##A",
            },
            "_numberFormatter": NumberFormatter {
              "_prepared": Object {
                "negative": false,
                "positive": Object {
                  "abbreviate": true,
                  "d": ".",
                  "decTemp": ".",
                  "groupTemp": "",
                  "isFunctional": false,
                  "maxPrecision": 2,
                  "numericPattern": "0.##",
                  "numericRegex": /,\\|\\\\\\./g,
                  "pattern": "0.##",
                  "percentage": false,
                  "postfix": "",
                  "prefix": "",
                  "t": ",",
                  "temp": 2,
                },
              },
              "abbreviations": Object {
                "-12": "p",
                "-15": "f",
                "-18": "a",
                "-21": "z",
                "-24": "y",
                "-3": "m",
                "-6": "Î¼",
                "-9": "n",
                "12": "T",
                "15": "P",
                "18": "E",
                "21": "Z",
                "24": "Y",
                "3": "k",
                "6": "M",
                "9": "G",
              },
              "decimalDelimiter": ".",
              "localeInfo": "",
              "pattern": "0.##A",
              "thousandDelimiter": "",
              "type": "U",
            },
            "_prepared": Object {
              "d": undefined,
              "pattern": "0.##A",
              "t": undefined,
            },
            "decimalDelimiter": ".",
            "localeInfo": Object {},
            "pattern": "0.##A",
            "thousandDelimiter": "",
            "type": "U",
            "values": Object {
              "max": 79336,
              "min": 6422,
              "range": 72914,
            },
          },
        ],
        "getContrastingColorTo": [Function],
        "headerColor": "#F2F2F2",
        "labels": Object {
          "auto": true,
          "headers": true,
          "leaves": true,
          "overlay": true,
          "values": true,
        },
        "level": 0,
        "showHeaders": false,
      },
      "type": "treemap",
    },
    Object {
      "key": "overlay",
      "layout": Object {
        "displayOrder": 10,
      },
      "settings": Object {
        "sources": Array [
          Object {
            "component": "treemap",
            "selector": "rect",
            "strategy": Object {
              "settings": Object {
                "align": 0.5,
                "fontFamily": "arial",
                "fontSize": 24,
                "justify": 0.5,
                "labels": Array [
                  Object {
                    "fill": [Function],
                    "label": [Function],
                    "wordBreak": "break-word",
                  },
                ],
                "padding": 8,
              },
              "type": "rows",
            },
          },
        ],
      },
      "type": "labels",
    },
    Object {
      "key": "tooltip",
      "settings": Object {
        "content": [Function],
        "duration": 2500,
        "extract": [Function],
        "placement": Object {
          "area": "target",
          "type": "bounds",
        },
      },
      "type": "tooltip",
    },
    Object {
      "key": "lasso",
      "settings": Object {
        "brush": Object {
          "components": Array [
            Object {
              "action": "add",
              "contexts": Array [
                "lassoContext",
              ],
              "data": Array [
                "select",
              ],
              "key": "treemap",
            },
          ],
        },
      },
      "type": "brush-lasso",
    },
  ],
  "interactions": Array [
    Object {
      "events": Object {
        "mouseleave": [Function],
        "mousemove": [Function],
      },
      "key": "native",
      "type": "native",
    },
    Object {
      "gestures": Array [
        Object {
          "events": Object {
            "pan": [Function],
            "panend": [Function],
            "panstart": [Function],
          },
          "type": "Pan",
        },
      ],
      "key": "hammer",
      "type": "hammer",
    },
  ],
  "scales": Object {
    "color": Object {
      "data": Object {
        "extract": Object {
          "field": "qMeasureInfo/0",
        },
      },
      "range": Array [
        "#c4cfda",
      ],
      "type": "color",
    },
  },
}
`;
